evalml/utils/nullable_type_utils.py

```diff
@@ -0,0 +1,103 @@
+import woodwork as ww
+
+
+def _downcast_nullable_X(X, handle_boolean_nullable=True, handle_integer_nullable=True):
+    """Removes Pandas nullable integer and nullable boolean dtypes from data by transforming
+        to other dtypes via Woodwork logical type transformations.
+
+    Args:
+        X (pd.DataFrame): Input data of shape [n_samples, n_features] whose nullable types will be changed.
+        handle_boolean_nullable (bool, optional): Whether or not to downcast data with BooleanNullable logical types.
+        handle_integer_nullable (bool, optional): Whether or not to downcast data with IntegerNullable or AgeNullable logical types.
+
+
+    Returns:
+        X with any incompatible nullable types downcasted to compatible equivalents.
+    """
+    if X.ww.schema is None:
+        X.ww.init()
+
+    select_param = []
+    if handle_boolean_nullable:
+        select_param.append("BooleanNullable")
+    if handle_integer_nullable:
+        select_param.append("IntegerNullable")
+        select_param.append("AgeNullable")
+
+    if not select_param:
+        return X
+
+    nullable_X_to_downcast = X.ww.select(select_param)
+    if not len(nullable_X_to_downcast.columns):
+        return X
+
+    cols_with_nans = set(
+        nullable_X_to_downcast.columns[nullable_X_to_downcast.isnull().any()],
+    )
+
+    new_ltypes = {
+        col: _get_downcast_logical_type(ltype, col in cols_with_nans)
+        for col, ltype in nullable_X_to_downcast.ww.logical_types.items()
+    }
```

Comment:
I'm having a bit of trouble understanding what's happening here. Are we leaving nullable types in the dataframe if there are no nans? Is that the desired behavior?

Reply:
Fair question -- I'll take a look and try to refactor to have this be easier to read.

So we leave nullable types in the dataframe if they are compatible (what `handle_boolean_nullable` and `handle_integer_nullable` decide).

We then gather the incompatible types into `nullable_X_to_downcast` and decide what logical type to convert them to based off of whether any nans are present (`_get_downcast_logical_type`). So for `IntegerNullable`, they can become the non nullable `Integer` if no nans are present but have to become `Double` if nans are present. Either way, though, we're removing all `IntegerNullable` uses.

Reply:
I attempted to rename things better and simplify some of the logic in https://github.com/alteryx/evalml/pull/4024/commits/601535b3826ff0c85592daf9722640639657f688 that will hopefully make the utils easier to understand. Let me know if anything still feels confusing